bin/mail/2016-03-29-ny-cs4all/generate-list.rb

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env ruby
+require_relative '../mailing-common/mailing-list-utils'
+require 'active_support'
+require 'active_support/core_ext/object/blank'
+require 'set'
+
+# Gather all email addresses *@schools.nyc.gov who are (any of):
+# A. Code Studio teachers who have already been through the K-5 PD.
+# B. Code Studio teachers who have not been through any PD.
+# C. Hour of Code teachers who aren't on Code Studio.
+
+def query_dashboard_teachers(query)
+  {}.tap do |results|
+    DASHBOARD_DB.fetch(query).each do |teacher|
+      email = teacher[:email]
+      results[email] = {email: email, name: teacher[:name], id: teacher[:id]}
+    end
+  end
+end
+
+# A. Query Code Studio teachers who have already been through the K-5 PD.
+def query_k5_pd_teachers
+  section_ids = Set.new
+  DB[:forms].where(kind: 'ProfessionalDevelopmentWorkshop').map(:data).each do |form_data|
+    section_id = JSON.parse(form_data)['section_id_s']
+    section_ids.add(section_id.to_i) unless section_id.blank?
+  end
+  puts "#{section_ids.count} K5 workshop section Ids loaded."
+
+  k5_pd_teachers = query_dashboard_teachers %Q(
+    SELECT DISTINCT culled_users.id, culled_users.email, culled_users.name
+    FROM (
+      SELECT * FROM users WHERE email like '%@schools.nyc.gov'
+    ) AS culled_users
+    JOIN followers ON (followers.student_user_id = culled_users.id)
+    JOIN sections ON (sections.id = followers.section_id)
+    WHERE sections.id in (#{section_ids.to_a.join(',')})
```

Comment:
Correct. [Sequel.fetch](http://sequel.jeremyevans.net/rdoc/classes/Sequel/Database.html#method-i-fetch) takes care of that.